ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,376 @@
+                    ListFoldingType::Prod(_) => Expression::Mul(Box::new(acc), Box::new(expr)),
+                };
+            }
+            Ok(acc)
+        }
+    }
+}
+
+/// Validates and returns the length of a list comprehension. Checks that the length of all iterables
+/// in the list comprehension is the same and returns an error if it's not.
+fn get_num_iterations(
+    lc: &ListComprehension,
+    symbol_table: &SymbolTable,
+) -> Result<usize, SemanticError> {
+    let lc_len = get_iterable_len(symbol_table, &lc.context()[0].1)?;
+    for (_, iterable) in lc.context().iter().skip(1) {
+        let iterable_len = get_iterable_len(symbol_table, iterable)?;
+        if iterable_len != lc_len {
+            return Err(SemanticError::InvalidListComprehension(
+                "All iterables in a list comprehension must have the same length".to_string(),
+            ));
+        }
+    }
+    Ok(lc_len)
+}
+
+/// Returns the length of an iterable.
+///
+/// # Errors
+/// - Returns an error if the iterable identifier is anything other than a vector in the symbol
+///   table if it's a variable.
+/// - Returns an error if the iterable is not of type in set:
+///   { IntegrityVariable, PublicInput, TraceColumns }
+fn get_iterable_len(
+    symbol_table: &SymbolTable,
+    iterable: &Iterable,
+) -> Result<usize, SemanticError> {
+    match iterable {
+        Iterable::Identifier(ident) => {
+            let ident_type = symbol_table.get_type(ident.name())?;
+            match ident_type {
+                IdentifierType::IntegrityVariable(var_type) => match var_type.value() {
+                    VariableType::Vector(vector) => Ok(vector.len()),
+                    _ => Err(SemanticError::InvalidListComprehension(
+                        "List comprehensions only allowed for vector variables.".to_string(),
+                    )),
+                },
+                IdentifierType::PublicInput(size) => Ok(*size),
+                IdentifierType::TraceColumns(trace_columns) => Ok(trace_columns.size()),
+                _ => Err(SemanticError::InvalidListComprehension(
+                    "IdentifierType {ident_type} not supported for list comprehensions".to_string(),
+                )),
+            }
+        }
+        Iterable::Range(range) | Iterable::Slice(_, range) => Ok(range.end() - range.start()),
+    }
+}
+
+/// Checks if the access index is valid. Returns an error if the access index is greater than
+/// the size of the vector.
+fn validate_access(i: usize, size: usize) -> Result<(), SemanticError> {
+    if i < size {
+        Ok(())
+    } else {
+        Err(SemanticError::InvalidListComprehension(
+            "Invalid access index".to_string(),
+        ))
+    }
+}
+
+/// Builds an [IterableContext] from a given list comprehension.
+fn build_iterable_context(lc: &ListComprehension) -> Result<IterableContext, SemanticError> {
```

Comment:
There's no documentation of `IterableContext` or of `Iterable`, and I think this is starting to get a bit hard to understand. Can you add brief docs for those so that this doc comment will be more helpful?